lib/fluent/plugin/gcloud_pubsub/client.rb

```diff
@@ -44,15 +74,54 @@ def topic(topic_name)
         client
       end
 
-      def publish(topic_name, messages)
-        topic(topic_name).publish do |batch|
-          messages.each do |m|
-            batch.publish m.message, m.attributes
+      def publish(topic_name, messages, compress_batches = false)
+        if compress_batches
+          topic(topic_name).publish(*compress_messages_with_zlib(messages, topic_name))
+        else
+          topic(topic_name).publish do |batch|
+            messages.each do |m|
+              batch.publish m.message, m.attributes
+            end
           end
         end
       rescue Google::Cloud::UnavailableError, Google::Cloud::DeadlineExceededError, Google::Cloud::InternalError => e
         raise RetryableError, "Google api returns error:#{e.class} message:#{e}"
       end
+
+      private
+
+      def compress_messages_with_zlib(messages, topic_name)
+        original_size = messages.sum(&:bytesize)
+        # This should never happen, only a programming error or major
+        # misconfiguration should lead to this situation. But checking against
+        # it here avoids a potential division by zero later on.
+        raise ArgumentError, "not compressing empty inputs" if original_size.zero?
+
+        # Here we're implicitly dropping the 'attributes' field of the messages
+        # that we're iterating over.
+        # This is fine, because the :attribute_keys config param is not
+        # supported when in compressed mode, so this field will always be
+        # empty.
+        packed_messages = messages.map(&:message).join(BATCHED_RECORD_SEPARATOR)
+
+        duration, compressed_messages = Fluent::GcloudPubSub::Metrics.measure_duration do
+          Zlib::Deflate.deflate(packed_messages)
+        end
+
+        @compression_duration.observe(
+          { topic: topic_name, algorithm: COMPRESSION_ALGORITHM_ZLIB },
+          duration,
+        )
+
+        compressed_size = compressed_messages.bytesize
+        @compression_ratio.observe(
```

Comment:
This to me looks like it is calculating the data-rate saving and not the ratio, which is what we want to do to ensue we can put this in known buckets of 0-1.

compression ratio = (original size)/(compressed size)
data-rate saving = 1 - (compressed size)/(uncompressed size)

Reply:
Ah yeah you're totally right I think. I was a bit torn between producing a number which would make sense on a dashboard, Prometheus' best-practice definition of a percentage/ratio, and avoiding potential division by zero; and ended up with something slightly confusing here.

I think you're right that we probably do want to keep this formula, as if we computed the actual compression ratio then the value of that is harder to predict (e.g. 10MiB/1KiB = 1024) and therefore define buckets for.

---
So for what we should replace it with:

I'd argue that we're not talking about 'data rate' here, as we're operating on a single blob rather than an indefinite stream, so it should be expressed as 'space savings' right?

- Current: 
  `fluentd_output_gcloud_pubsub_messages_compression_ratio`

  Variable name: `@compression_ratio`
- Proposal 1: 
  `fluentd_output_gcloud_pubsub_messages_compression_space_savings_ratio`
  
  Variable name: `@compression_space_saving`

  I'd like to keep `compression` in the name as it groups it with the `compression_duration_seconds` metric.
- Proposal 2:
 `fluentd_output_gcloud_pubsub_messages_compressed_size_per_original_size_ratio`

  Variable name: `@compression_ratio` ??

  This is closer to the [Prometheus suggestion](https://prometheus.io/docs/practices/naming/#base-units), but feels a little bit awkward?

Reply:
That naming of Proposal 2 is really doing it for me 😆 

Lets do that!